File: apps/agent/src/utils/pg-store.ts
import { AsyncBatchedStore, BaseStore } from "@langchain/langgraph";
import type {
  Operation,
  OperationResults,
  GetOperation,
  PutOperation
} from "@langchain/langgraph";
import { Pool } from "pg";

// Type guards for Operation types
function isGetOperation(op: Operation): op is GetOperation {
    return (op as GetOperation).namespace !== undefined && (op as GetOperation).key !== undefined && (op as any).value === undefined;
}
function isPutOperation(op: Operation): op is PutOperation {
    return (op as PutOperation).namespace !== undefined && (op as PutOperation).key !== undefined && (op as PutOperation).value !== undefined;
}
// function isDeleteOperation(op: Operation): op is DeleteOperation {
//     return (op as DeleteOperation).namespace !== undefined && (op as DeleteOperation).key !== undefined && (op as any).value === undefined && (op as any).index === undefined;
// }

export class PostgresStore extends BaseStore {
    private pool: Pool;
    private tableName: string;

    constructor(connString: string, tableName = "langgraph_pg_store") {
        super();
        this.pool = new Pool({ connectionString: connString });
        this.tableName = tableName;
    }

    // Helper to create composite key
    private makeKey(namespace: string[], key: string): string {
        return namespace.length ? `${namespace.join(":")}:${key}` : key;
    }

    // BaseStore methods
    async batch<Op extends Operation[]>(operations: Op): Promise<OperationResults<Op>> {
        const results: Partial<OperationResults<Op>> = {};
        for (let i = 0; i < operations.length; i++) {
            const op = operations[i];
            // Type guards for operation types
            if (isGetOperation(op)) {
                results[i] = await this.get(op.namespace, op.key);
            } else if (isPutOperation(op)) {
                await this.put(op.namespace, op.key, op.value || {}, op.index);
                results[i] = undefined;
            // } else if (isDeleteOperation(op)) {
            //     await this.delete(op.namespace, op.key);
            //     results[i] = undefined;
            } else {
                results[i] = undefined;
            }
        }
        return results as OperationResults<Op>;
    }


    async get(namespace: string[], key: string): Promise<any> {
        const compositeKey = this.makeKey(namespace, key);
        const result = await this.batchLoad([compositeKey]);
        return result[compositeKey] ?? null;
    }

    async put(namespace: string[], key: string, value: Record<string, any>, index?: false | string[]): Promise<void> {
        const compositeKey = this.makeKey(namespace, key);
        await this.batchSave({ [compositeKey]: value });
    }

    async delete(namespace: string[], key: string): Promise<void> {
        const compositeKey = this.makeKey(namespace, key);
        await this.pool.query(
            `DELETE FROM ${this.tableName} WHERE key = $1`,
            [compositeKey]
        );
    }

    async getAll(namespace: string[]): Promise<Record<string, any>> {
        const prefix = namespace.length ? `${namespace.join(":")}:` : '';
        const { rows } = await this.pool.query(
            `SELECT key, value FROM ${this.tableName} WHERE key LIKE $1`,
            [`${prefix}%`]
        );
        const result: Record<string, any> = {};
        for (const row of rows) {
            result[row.key] = row.value;
        }
        return result;
    }

    async clear(namespace: string[]): Promise<void> {
        const prefix = namespace.length ? `${namespace.join(":")}:` : '';
        await this.pool.query(
            `DELETE FROM ${this.tableName} WHERE key LIKE $1`,
            [`${prefix}%`]
        );
    }

    async search(namespacePrefix: string[], options?: { filter?: Record<string, any>, limit?: number, offset?: number, query?: string }): Promise<{ key: string; value: any; namespace: string[]; createdAt: Date; updatedAt: Date }[]> {
        const prefix = namespacePrefix.length ? `${namespacePrefix.join(":")}:` : '';
        let queryStr = `SELECT key, value FROM ${this.tableName} WHERE key LIKE $1`;
        const params: any[] = [`${prefix}%`];
        if (options?.query) {
            queryStr += ` AND (key ILIKE $2 OR value::text ILIKE $2)`;
            params.push(`%${options.query}%`);
        }
        if (options?.limit) {
            queryStr += ` LIMIT $${params.length + 1}`;
            params.push(options.limit);
        }
        if (options?.offset) {
            queryStr += ` OFFSET $${params.length + 1}`;
            params.push(options.offset);
        }
        const { rows } = await this.pool.query(queryStr, params);
        return rows.map((row: any) => ({ key: row.key, value: row.value, namespace: namespacePrefix, createdAt: row.createdAt, updatedAt: row.updatedAt }));
    }

    async initialize() {
        await this.pool.query('SET search_path TO langgraph');
        await this.pool.query(`
        CREATE TABLE IF NOT EXISTS ${this.tableName} (
          key TEXT PRIMARY KEY,
          value JSONB
        )
      `);
    }

    async batchLoad(keys: string[]): Promise<Record<string, any>> {
        if (!keys.length) return {};
        const { rows } = await this.pool.query(
            `SELECT key, value FROM ${this.tableName} WHERE key = ANY($1::text[])`,
            [keys]
        );
        const result: Record<string, any> = {};
        for (const row of rows) {
            result[row.key] = row.value;
        }
        return result;
    }

    async batchSave(pairs: Record<string, any>): Promise<void> {
        if (!Object.keys(pairs).length) return;
        const client = await this.pool.connect();
        try {
            await client.query('BEGIN');
            for (const [key, value] of Object.entries(pairs)) {
                await client.query(
                    `INSERT INTO ${this.tableName} (key, value) VALUES ($1, $2)
                    ON CONFLICT (key) DO UPDATE SET value = EXCLUDED.value`,
                    [key, value]
                );
            }
            await client.query('COMMIT');
        } catch (e) {
            await client.query('ROLLBACK');
            throw e;
        } finally {
            client.release();
        }
    }
}

File: apps/agent-python/src/utils/pg_store.py
import os
import json
import psycopg
from psycopg_pool import ConnectionPool
from typing import Dict, List, Any, Optional
from langgraph.store.base import BaseStore


class PostgresStore(BaseStore):
    """PostgreSQL-backed store for LangGraph."""

    def __init__(self, conn_string: str, table_name: str = "langgraph_pg_store"):
        """
        Initialize PostgreSQL store.

        Args:
            conn_string: PostgreSQL connection string
            table_name: Name of the table to use for storage
        """
        super().__init__()
        self.pool = ConnectionPool(conn_string, min_size=1, max_size=10)
        self.table_name = table_name

    def _make_key(self, namespace: List[str], key: str) -> str:
        """Create composite key from namespace and key."""
        if namespace:
            return f"{':'.join(namespace)}:{key}"
        return key

    def get(self, namespace: List[str], key: str) -> Optional[Any]:
        """Get a value by namespace and key."""
        composite_key = self._make_key(namespace, key)
        result = self.batch_load([composite_key])
        return result.get(composite_key)

    def put(
        self,
        namespace: List[str],
        key: str,
        value: Dict[str, Any],
        index: Optional[List[str]] = None
    ) -> None:
        """Put a value in the store."""
        composite_key = self._make_key(namespace, key)
        self.batch_save({composite_key: value})

    def delete(self, namespace: List[str], key: str) -> None:
        """Delete a value from the store."""
        composite_key = self._make_key(namespace, key)
        with self.pool.connection() as conn:
            with conn.cursor() as cursor:
                cursor.execute(
                    f"DELETE FROM {self.table_name} WHERE key = %s",
                    (composite_key,)
                )
                conn.commit()

    def get_all(self, namespace: List[str]) -> Dict[str, Any]:
        """Get all values in a namespace."""
        prefix = f"{':'.join(namespace)}:" if namespace else ''
        with self.pool.connection() as conn:
            with conn.cursor() as cursor:
                cursor.execute(
                    f"SELECT key, value FROM {self.table_name} WHERE key LIKE %s",
                    (f"{prefix}%",)
                )
                rows = cursor.fetchall()
                result = {}
                for row in rows:
                    result[row['key']] = row['value']
                return result

    def clear(self, namespace: List[str]) -> None:
        """Clear all values in a namespace."""
        prefix = f"{':'.join(namespace)}:" if namespace else ''
        with self.pool.connection() as conn:
            with conn.cursor() as cursor:
                cursor.execute(
                    f"DELETE FROM {self.table_name} WHERE key LIKE %s",
                    (f"{prefix}%",)
                )
                conn.commit()

    def search(
        self,
        namespace_prefix: List[str],
        filter_dict: Optional[Dict[str, Any]] = None,
        limit: Optional[int] = None,
        offset: Optional[int] = None,
        query: Optional[str] = None
    ) -> List[Dict[str, Any]]:
        """Search for values in the store."""
        prefix = f"{':'.join(namespace_prefix)}:" if namespace_prefix else ''
        query_str = f"SELECT key, value FROM {self.table_name} WHERE key LIKE %s"
        params = [f"{prefix}%"]

        if query:
            query_str += " AND (key ILIKE %s OR value::text ILIKE %s)"
            params.extend([f"%{query}%", f"%{query}%"])

        if limit:
            query_str += f" LIMIT %s"
            params.append(limit)

        if offset:
            query_str += f" OFFSET %s"
            params.append(offset)

        with self.pool.connection() as conn:
            with conn.cursor() as cursor:
                cursor.execute(query_str, params)
                rows = cursor.fetchall()
                return [
                    {
                        'key': row['key'],
                        'value': row['value'],
                        'namespace': namespace_prefix
                    }
                    for row in rows
                ]

    def initialize(self) -> None:
        """Initialize the store (create table if not exists)."""
        with self.pool.connection() as conn:
            with conn.cursor() as cursor:
                cursor.execute("SET search_path TO langgraph")
                cursor.execute(f"""
                    CREATE TABLE IF NOT EXISTS {self.table_name} (
                        key TEXT PRIMARY KEY,
                        value JSONB
                    )
                """)
                conn.commit()

    def batch_load(self, keys: List[str]) -> Dict[str, Any]:
        """Load multiple keys at once."""
        if not keys:
            return {}

        with self.pool.connection() as conn:
            with conn.cursor() as cursor:
                cursor.execute(
                    f"SELECT key, value FROM {self.table_name} WHERE key = ANY(%s)",
                    (keys,)
                )
                rows = cursor.fetchall()
                result = {}
                for row in rows:
                    result[row['key']] = row['value']
                return result

    def batch_save(self, pairs: Dict[str, Any]) -> None:
        """Save multiple key-value pairs at once."""
        if not pairs:
            return

        with self.pool.connection() as conn:
            with conn.cursor() as cursor:
                for key, value in pairs.items():
                    # Convert value to JSON if it's a dict
                    json_value = json.dumps(value) if isinstance(value, dict) else value
                    cursor.execute(f"""
                        INSERT INTO {self.table_name} (key, value)
                        VALUES (%s, %s)
                        ON CONFLICT (key) DO UPDATE SET value = EXCLUDED.value
                    """, (key, json_value))
                conn.commit()

    def batch(self, operations: List[Any]) -> List[Any]:
        """Execute a batch of operations synchronously."""
        results = []
        for op in operations:
            if hasattr(op, 'namespace') and hasattr(op, 'key'):
                if hasattr(op, 'value'):
                    # Put operation
                    self.put(op.namespace, op.key, op.value)
                    results.append(None)
                else:
                    # Get operation
                    result = self.get(op.namespace, op.key)
                    results.append(result)
            else:
                results.append(None)
        return results

    async def abatch(self, operations: List[Any]) -> List[Any]:
        """Execute a batch of operations asynchronously."""
        # For now, just call the sync version
        # In a production system, you'd want true async operations
        return self.batch(operations)

    def close(self) -> None:
        """Close all connections in the pool."""
        self.pool.close()
